cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,888 @@
+   {
+      char *local_file;
+      bool need_download = false;
+      cloud_storage_item_t *current_metadata;
+
+      if (remote_file->last_sync_time < time(NULL) - 30)
+      {
+         current_metadata = _provider->get_file_metadata(remote_file);
+         if (!current_metadata)
+         {
+            continue;
+         }
+         _update_existing_item(remote_file, current_metadata);
+         cloud_storage_item_free(current_metadata);
+      }
+
+      local_file = _get_absolute_filename(folder_type, remote_file->name);
+
+      if (!local_file)
+      {
+         need_download = false;
+      } else if (_have_local_file_for_remote_file(folder_type, remote_file))
+      {
+         need_download = _file_need_download(local_file, remote_file);
+      } else
+      {
+         need_download = true;
+      }
+
+      if (need_download)
+      {
+         _download_file(local_file, remote_file);
+      }
+   }
+}
+
+static bool _prepare_folder(folder_type_t folder_type)
+{
+   const char *folder_name;
+   cloud_storage_item_t **folder;
+
+   switch (folder_type)
+   {
+      case CLOUD_STORAGE_GAME_STATES:
+         folder_name = GAME_STATES_FOLDER_NAME;
+         folder = &_game_states;
+         break;
+      default:
+         return false;
+   }
+
+   if (*folder == NULL)
+   {
+      *folder = _provider->get_folder_metadata(folder_name);
+      if (*folder)
+      {
+         _provider->list_files(*folder);
+         return true;
+      } else
+      {
+         *folder = _provider->create_folder(folder_name);
+         return *folder != NULL;
+      }
+   } else
+   {
+      return true;
+   }
+}
+
+static void _sync_files_execute(folder_type_t folder_type)
+{
+   if (!_prepare_folder(folder_type))
+   {
+      return;
+   }
+
+   _sync_files_upload(folder_type);
+   _sync_files_download(folder_type);
+}
+
+void cloud_storage_sync_files(void)
+{
+   bool sync_states;
+   global_t *global = global_get_ptr();
+   char *raw_folder;
+   int i;
+   struct _operation_queue_item *queue_item = NULL;
+   struct _operation_queue_item *previous_queue_item = NULL;
+
+   raw_folder = _config_value_to_directory(global->name.savestate);
+   sync_states = strcmp(_game_states_config_value, raw_folder) != 0;
+
+   if (!sync_states)
+   {
+      return;
+   } else
+   {
+      strcpy(_game_states_config_value, global->name.savestate);
+   }
+
+   slock_lock(_operation_queue_mutex);
+
+   if (!_provider->ready_for_request())
+   {
+      slock_unlock(_operation_queue_mutex);
+      return;
+   }
+
+   for (queue_item = _operation_queue; queue_item != NULL; queue_item = queue_item->next)
+   {
+      if (queue_item->operation == UPLOAD_FILE || queue_item->operation == DOWNLOAD_FILE)
+      {
+         if (!previous_queue_item)
+         {
+            _free_operation_queue_item(queue_item);
+         } else
+         {
+            struct _operation_queue_item *next_item;
+
+            next_item = queue_item->next;
```

Comment:
Here you are deleting the current item (queue_item) and at the end of the block, queue item points to the next item to the item you just deleted.
However the loop will execute `queue_item = queue_item->next` when processing the next iteration, therefore skipping an item in the queue to process. I'm pretty sure this is not on purpose right?
If you want to filter elements of a queue, it's much simpler to builld a new queue with the elements that you do *not* want to delete.
I would add that perhaps it would be a good idea to use a generic queue (not sure whether there's one in RA, but maybe we could create one! With tests :P) and call some generic function like delete_queue_items_filter(queue, evaluate_elem_delete_callback). Just to give you an idea.